control_plane/src/attachment_service.rs

```diff
@@ -220,9 +241,213 @@ impl AttachmentService {
             .expect("non-Unicode path")
     }
 
+    /// PIDFile for the postgres instance used to store attachment service state
+    fn postgres_pid_file(&self) -> Utf8PathBuf {
+        Utf8PathBuf::from_path_buf(
+            self.env
+                .base_data_dir
+                .join("attachment_service_postgres.pid"),
+        )
+        .expect("non-Unicode path")
+    }
+
+    /// In order to access database migrations, we need to find the Neon source tree
+    async fn find_source_root(&self) -> anyhow::Result<Utf8PathBuf> {
+        // We assume that either prd or our binary is in the source tree. The former is usually
+        // true for automated test runners, the latter is usually true for developer workstations. Often
+        // both are true, which is fine.
+        let candidate_start_points: Vec<Utf8PathBuf> = vec![
+            // Current working directory
+            Utf8PathBuf::from_path_buf(std::env::current_dir()?).unwrap(),
+            // Directory containing the binary we're running inside
+            Utf8PathBuf::from_path_buf(env::current_exe()?.parent().unwrap().to_owned()).unwrap(),
+            // Source dir on github action workers
+            Utf8PathBuf::from_str("/__w/neon").unwrap(),
+        ];
+
+        // For each candidate start point, search through ancestors looking for a neon.git source tree root
+        for start_point in &candidate_start_points {
+            // Start from the build dir: assumes we are running out of a built neon source tree
+            let mut cursor = start_point.clone();
+            loop {
+                // A crude approximation: the root of the source tree is whatever contains a "control_plane"
+                // subdirectory.
+                let control_plane = cursor.join("control_plane");
+                if tokio::fs::try_exists(&control_plane).await? {
+                    return Ok(cursor);
+                } else {
+                    cursor = match cursor.parent() {
+                        Some(d) => d.to_owned(),
+                        None => {
+                            // Give up searching from this starting point, try the next one
+                            break;
+                        }
+                    };
+                }
+            }
+        }
+
+        // Fall-through
+        Err(anyhow::anyhow!(
+            "Could not find control_plane src dir, after searching ancestors of {candidate_start_points:?}"
+        ))
+    }
+
+    /// Find the directory containing postgres binaries, such as `initdb` and `pg_ctl`
+    ///
+    /// This usually uses ATTACHMENT_SERVICE_POSTGRES_VERSION of postgres, but will fall back
+    /// to other versions if that one isn't found.  Some automated tests create circumstances
+    /// where only one version is available in pg_distrib_dir, such as `test_remote_extensions`.
+    pub async fn get_pg_bin_path(&self) -> anyhow::Result<PathBuf> {
```

Comment:
```suggestion
    pub async fn get_pg_bin_dir(&self) -> anyhow::Result<PathBuf> {
```

I know that in unix, `*PATH` env variables are an established concept pointing to directories, but it is still clearer to have directory in the name directly.

Reply:
In this case I agree it's clearer to say "bin_dir", so that reader knows we're talking about the path to the directory, rather than the path to a particular binary.  Changed.